### PR TITLE
♻️ Trim `dd/Package.hpp`

### PR DESCRIPTION
**Pull request by burgholzer**
## Description

This PR brings some more structure to the DD package by factoring out methods that do not necessarily need any package internal methods/objects.
This creates a better separation of concern and should speed up build times.
Overall, this PR eliminates >400 LoC from the `dd/Package.hpp` file and only results in a total of +3 LoC for the actual library code.
The remaining additions are tests and documentation.

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: include/dd/Complex.hpp [hunk 1]
```diff
@@ -102,6 +103,12 @@ struct Complex {
    * @see CTEntry::writeBinary
    */
   void writeBinary(std::ostream& os) const;
+
+  /**
+   * @brief Convert the Complex number to an std::complex<fp>.
+   * @returns The std::complex<fp> representation of the Complex number.
+   */
+  explicit operator std::complex<fp>() const noexcept;
```
Comment (hillmich) [inline include/dd/Complex.hpp]: Could this be `[[nodiscard]]`?

File: include/dd/Edge.hpp [hunk 2]
```diff
@@ -1,14 +1,50 @@
...
+                     bool>;
+
+using AmplitudeFunc =
+    std::function<void(const std::size_t, const std::complex<fp>&)>;
+using ProbabilityFunc = std::function<void(const std::size_t, const fp&)>;
+using MatrixEntryFunc = std::function<void(const std::size_t, const std::size_t,
+                                           const std::complex<fp>&)>;
+
+/**
+ * @brief A DD node with an edge weight
```
Comment (hillmich) [inline include/dd/Edge.hpp]: Maybe describe as follows?

> A weighted edge pointing to a DD node.

File: include/dd/Edge.hpp [hunk 3]
```diff
@@ -1,14 +1,50 @@
...
+template <typename T>
+using isVector = std::enable_if_t<std::is_same_v<T, vNode>, bool>;
+template <typename T>
+using isMatrix = std::enable_if_t<std::is_same_v<T, mNode>, bool>;
+template <typename T>
+using isDensityMatrix = std::enable_if_t<std::is_same_v<T, dNode>, bool>;
+template <typename T>
+using isMatrixVariant =
+    std::enable_if_t<std::is_same_v<T, mNode> || std::is_same_v<T, dNode>,
+                     bool>;
```
Comment (hillmich) [inline include/dd/Edge.hpp:33]: That's pretty neat :+1:

**Review by hillmich — DISMISSED** [review]
LGTM! Below are a few minor comments you can happily apply or ignore if ill-suited and then resolve.

File: include/dd/Edge.hpp (hunk 2 above)
Comment (burgholzer) [reply include/dd/Edge.hpp]: Good suggestion! Done in https://github.com/cda-tum/mqt-core/pull/434/commits/a2964ae7e6cecd95314d3fa9c2e0402bbf68c3ec

File: include/dd/Complex.hpp (hunk 1 above)
Comment (burgholzer) [reply include/dd/Complex.hpp]: Sure can. Good catch. Done in https://github.com/cda-tum/mqt-core/pull/434/commits/84762700ab4fadf47c8fdac53277d3f0aa57c193

**Comment by burgholzer** [comment]
> LGTM! Below are a few minor comments you can happily apply or ignore if ill-suited and then resolve.

Thanks for your time and the suggestions!
I enabled auto-merge.
Once you resolve the conversations and (re-)approve the PR, this should be merged automatically.

**Review by hillmich — APPROVED** [review]
LGTM!

**Pull request merged**